core/test/value_and_variance_test.cpp

```diff
@@ -98,6 +98,108 @@ TEST(ValueAndVarianceTest, binary_divide_equals) {
             lhs.variance);
 }
 
+TEST(ValueAndVarianceTest, comparison) {
+  ValueAndVariance a1{1.0, 2.0};
+  ValueAndVariance a2{1.0, 3.0}; // same as a1 but difference variance
+  ValueAndVariance b{2.0, 2.0};
+
+  EXPECT_TRUE(a1 == a1);
+  EXPECT_TRUE(a1 == a2);
+  EXPECT_FALSE(a1 == b);
+
+  EXPECT_FALSE(a1 != a1);
+  EXPECT_FALSE(a1 != a2);
+  EXPECT_TRUE(a1 != b);
+
+  EXPECT_FALSE(a1 < a1);
+  EXPECT_FALSE(a1 < a2);
+  EXPECT_TRUE(a1 < b);
+  EXPECT_FALSE(b < a1);
+
+  EXPECT_FALSE(a1 > a1);
+  EXPECT_FALSE(a1 > a2);
+  EXPECT_FALSE(a1 > b);
+  EXPECT_TRUE(b > a1);
+
+  EXPECT_TRUE(a1 <= a1);
+  EXPECT_TRUE(a1 <= a2);
+  EXPECT_TRUE(a1 <= b);
+  EXPECT_FALSE(b <= a1);
+
+  EXPECT_TRUE(a1 >= a1);
+  EXPECT_TRUE(a1 >= a2);
+  EXPECT_FALSE(a1 >= b);
+  EXPECT_TRUE(b >= a1);
+}
+
+TEST(ValueAndVarianceTest, comparison_no_variance_lhs) {
+  ValueAndVariance a1{1.0, 2.0};
+  ValueAndVariance a2{1.0, 3.0}; // same as a1 but difference variance
+  ValueAndVariance b{2.0, 2.0};
+
+  EXPECT_TRUE(a1.value == a1);
+  EXPECT_TRUE(a1.value == a2);
+  EXPECT_FALSE(a1.value == b);
+
+  EXPECT_FALSE(a1.value != a1);
+  EXPECT_FALSE(a1.value != a2);
+  EXPECT_TRUE(a1.value != b);
+
+  EXPECT_FALSE(a1.value < a1);
+  EXPECT_FALSE(a1.value < a2);
+  EXPECT_TRUE(a1.value < b);
+  EXPECT_FALSE(b.value < a1);
+
+  EXPECT_FALSE(a1.value > a1);
+  EXPECT_FALSE(a1.value > a2);
+  EXPECT_FALSE(a1.value > b);
+  EXPECT_TRUE(b.value > a1);
+
+  EXPECT_TRUE(a1.value <= a1);
+  EXPECT_TRUE(a1.value <= a2);
+  EXPECT_TRUE(a1.value <= b);
+  EXPECT_FALSE(b.value <= a1);
+
+  EXPECT_TRUE(a1.value >= a1);
+  EXPECT_TRUE(a1.value >= a2);
+  EXPECT_FALSE(a1.value >= b);
+  EXPECT_TRUE(b.value >= a1);
+}
+
+TEST(ValueAndVarianceTest, comparison_no_variance_rhs) {
+  ValueAndVariance a1{1.0, 2.0};
+  ValueAndVariance a2{1.0, 3.0}; // same as a1 but difference variance
+  ValueAndVariance b{2.0, 2.0};
+
+  EXPECT_TRUE(a1 == a1.value);
+  EXPECT_TRUE(a1 == a2.value);
+  EXPECT_FALSE(a1 == b.value);
+
+  EXPECT_FALSE(a1 != a1.value);
+  EXPECT_FALSE(a1 != a2.value);
+  EXPECT_TRUE(a1 != b.value);
+
+  EXPECT_FALSE(a1 < a1.value);
+  EXPECT_FALSE(a1 < a2.value);
+  EXPECT_TRUE(a1 < b.value);
+  EXPECT_FALSE(b < a1.value);
+
+  EXPECT_FALSE(a1 > a1.value);
+  EXPECT_FALSE(a1 > a2.value);
+  EXPECT_FALSE(a1 > b.value);
+  EXPECT_TRUE(b > a1.value);
+
+  EXPECT_TRUE(a1 <= a1.value);
+  EXPECT_TRUE(a1 <= a2.value);
+  EXPECT_TRUE(a1 <= b.value);
+  EXPECT_FALSE(b <= a1.value);
+
+  EXPECT_TRUE(a1 >= a1.value);
+  EXPECT_TRUE(a1 >= a2.value);
+  EXPECT_FALSE(a1 >= b.value);
+  EXPECT_TRUE(b >= a1.value);
+}
+
```

Comment:
Would be nice if there was some code reuse between `comparison_no_variance_rhs` and `comparison_no_variance_lhs` - 30 lines each and look to be identical apart from the order of the operands. Maybe something like `testing::Combine` would be useful. Not crtical for this so maybe a consideration for a follow on step.

Reply:
I had tried to use a templated helper function, so I could pass, e.g., `a` in one case and `a.value` in another case, but actually what I needed was for this to be done to either the LHS or RHS in each of the operators, so that did not work.

Reply:
I'll merge this for now, if you have an idea on how to reduce the duplication in a non-complicating way I am happy to change later.